### PR TITLE
[Merged by Bors] - feat: port CategoryTheory.Limits.Shapes.Diagonal

### PR DESCRIPTION
**Pull request by joelriou**
---
- [x] depends on: #2871 

There is currently a deterministic timeout which can be circumvented by `set_option maxHeartbeats 400000`

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

**Comment by semorrison** [comment]
This PR/issue depends on:
* leanprover-community/mathlib4#2871
By **[Dependent Issues](https://github.com/z0al/dependent-issues)** (🤖). Happy coding!

**Comment by semorrison** [comment]
This PR/issue depends on:
* ~~leanprover-community/mathlib4#2871~~
By **[Dependent Issues](https://github.com/z0al/dependent-issues)** (🤖). Happy coding!

File: Mathlib/CategoryTheory/Limits/Shapes/Diagonal.lean [hunk 1]
```diff
@@ -0,0 +1,461 @@
...
+    pullback.lift (pullback.fst ≫ i₁ ≫ pullback.fst)
+      (pullback.map _ _ _ _ (𝟙 _) (𝟙 _) pullback.snd (Category.id_comp _).symm
+        (Category.id_comp _).symm)
+      (pullback.hom_ext
+        (by simp only [Category.assoc, diagonal_fst, Category.comp_id, limit.lift_π,
+          PullbackCone.mk_pt, PullbackCone.mk_π_app, limit.lift_π_assoc, cospan_left])
+        (by simp only [Category.assoc, diagonal_snd, Category.comp_id, pullback.lift_snd,
+            pullback.lift_snd_assoc, pullback.condition_assoc]))
+  hom_inv_id := by
+    apply pullback.hom_ext
```
Comment (ChrisHughes24) [inline Mathlib/CategoryTheory/Limits/Shapes/Diagonal.lean:155]: Can these be fixed by setting priorities on the `@[ext]` attributes. In mathlib3 I think there were no priorities and the most recent thing was used first, but in mathlib4 we should be setting priorities to make this work. You can write `@[ext 1100]` to increase the priority, and the default is `1000`.

**Comment by ChrisHughes24** [comment]
bors merge

**Pull request closed without merging**